packages/hooks/src/useMutationObserver.ts

```diff
@@ -0,0 +1,40 @@
+import { useEffect, useState } from 'react';
+
+type MutationHandler<Value> = (
+  mutations: Array<MutationRecord>,
+  observer: MutationObserver,
+) => Value;
+
+/**
+ * Hook to subscribe to changes on the DOM.
+ * @param target HTML element that is subscribed to DOM changes.
+ * @param options Object with information about what DOM changes to subscribe to.
+ * @param callback Callback function to execute inside of MutationObserver instance.
+ * @param enabled Determines whether or not the hook should run, defaults to true.
+ */
+export default function useMutationObserver<Value>(
+  target: HTMLElement | null,
+  options: MutationObserverInit,
+  callback: MutationHandler<Value>,
+  enabled: boolean = true,
+): Value | undefined {
+  const [value, setValue] = useState<Value>();
+
+  useEffect(() => {
+    if (!enabled) {
+      return;
+    }
+
+    const observer = new MutationObserver((...args) => {
+      setValue(callback(...args));
+    });
+
+    if (target) {
+      observer.observe(target, options);
+    }
+
+    return () => observer.disconnect();
```

Comment:
```js
return observer.disconnect
```

Reply:
I'd want to verify that `disconnect` is bound to the observer instance.

Reply:
Curious why this is relevant here? Having this being invoked within an arrow function shouldn't affect whether or not the method is bound to the instance, right?

Maybe I'm missing something here, but I was under the impression this change wouldn't fundamentally change how the function operates in this case. Are you afraid that it's setting something attached to the function scope that you want to persist when this is called?

Reply:
Who is calling `disconnect`? I think it'd be React? 

It's the same issue as event handlers.

If you do `addEventListener('click', observer.disconnect)` then disconnect is being called not in the context of `observer`. From the event listener instead.

Reply:
That's really annoying it's not bound to the instance... :/ Alright, fine as is I suppose then.